### PR TITLE
Template for framework-agnostic tests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

There are a few cross-framework pain points I often encounter: 
1. Ensuring the interface of `.generate()` or models stay consistent across frameworks;
2. Ensuring that TF doesn't get neglected and satisfies 1., when contributors add features/fixes on the PT side;
3. Blocking cases where the interface is the same, but there are numerical differences.

After a brief chat with @sgugger, I thought inheritable framework-agnostic tests could be nice to help with these problems. This week I'll have to write a few TF `.generate()` tests that already exist on the PT side, so this could be a great chance to kill 2 birds with one stone. 

However, I'd like to get the pattern right, hence this small PR -- replaces a pair of framework-specific tests with a framework-agnostic test. `Flax` is intentionally left out, as it is missing many testable features and is not being maintained, but it can easily be added in the future.

Let me know what you think of it!

### CODE REVIEW

File: utils/tests_fetcher.py [hunk 1]
```diff
@@ -466,12 +466,13 @@ def module_to_test_file(module_fname):
 # This list contains the list of test files we expect never to be launched from a change in a module/util. Those are
 # launched separately.
 EXPECTED_TEST_FILES_NEVER_TOUCHED = [
-    "tests/utils/test_doc_samples.py",  # Doc tests
+    "tests/generation/test_framework_agnostic.py",  # Mixins inherited by actual test classes
```
Comment (gante) [inline utils/tests_fetcher.py:469]: (added this line and sorted the others)

File: tests/generation/test_framework_agnostic.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+"""
+Framework agnostic tests for generate()-related methods.
```
Comment (gante) [inline tests/generation/test_framework_agnostic.py:2]: The goal would be to expand this file with other generation integration tests, logits processors tests, and model-inheritable generation tests

**Review by amyeroberts — APPROVED** [review]
Nice - LGTM! 

Only have a small question. For your third point: 
> 3. Blocking cases where the interface is the same, but there are numerical differences.

I might be missing something, but not clear to me how/where this is tested here.

**Comment by gante** [comment]
> I might be missing something, but not clear to me how/where this is tested here.

@amyeroberts I see, this example does not test numerical differences :) 

However, consider the tests [here](https://github.com/huggingface/transformers/blob/main/tests/generation/test_logits_process.py), which also exist for TF. They are primarily numerical tests, they test the resulting vector against a constant. It is not uncommon for a user to find an edge case on these processors, fix the PT side of it, and add a few more numerical checks. These new numerical checks would fail in TF, but I often can't convince the users to make the change on the TF side (and then fail to follow up myself after it is merged). Result: a bug that is trivial to fix in that moment gets lost 🙈 This unified framework would prevent it -- the fix would need to include the corresponding TF change to pass the tests (either fixed by the user or by one of us).

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: tests/generation/test_framework_agnostic.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import numpy as np
+
+from transformers import AutoTokenizer
+
+
+class GenerationIntegrationTestsMixin:
+
+    # To be populated by the child classes
+    framework_dependent_parameters = {
+        "AutoModelForSeq2SeqLM": None,
```
Comment (ydshieh) [inline tests/generation/test_framework_agnostic.py:14]: I kind feel this should be something like `model_class: None`, and in `model_class: explicit class or auto calss` in subclasses.

But as you mentioned, this file is meant to be expanded, there is no real need to figure out what would be the best name at this moment.

**Review by ydshieh — APPROVED** [review]
Thank you @gante ! I ❤️  unified testing!
